### DISCUSSION TITLE
RSVP Task - Participant response window longer than animation frame time

### DISCUSSION THREAD

**Discussion by william-hopper** [jspsych/jsPsych#747] [Q&A]
Hello!

Firstly, thank you very much for such a great JS library!

Now my question:
I am trying to implement a Rapid Serial Visual Presentation Task (RSVP) during which the on-screen stimuli will update every 350 ms. Participants need to respond (by pressing the space bar) only when they see the target stimuli within the flux, and they have 1000 ms to do so for a correct response. On one trial, there will be a total of 5 targets to respond to, for a total of 40 (including the targets) stimuli appearing on screen during the trial. As well as knowing when they respond correctly, I also need to know when they respond incorrectly: participants will only be allowed a certain number of so-called "false alarms" on a given trial to prevent them from simply continuously pressing the space bar throughout.

Does anyone have a suggestion on the best way to go about coding this? The asynchronous nature of the stimuli presentation and response window is giving me trouble!

Thanks!

**Comment by vijaymarupudi** [comment]
What is the type of stimuli you are using, would plain html be able to display it? If so, you could probably modify the html-keyboard-response plugin to suit your needs, otherwise, let's talk!

Just to clarify, each trial will contain 40 stimuli, and each stimuli will be displayed for 350ms? Would there be a gap between the stimuli to allow for the 1000ms response duration?

It might be useful to consider a 'jsPsych trial' for each stimulus within an 'experiment trial', there does not necessarily have to be a one to one correspondance. You can use [on_finish] callbacks to tag with accuracy data, and possibly [looping timelines] for validation.

I'll wait for more information before specifics!

[looping timelines]: https://www.jspsych.org/overview/timeline/#looping-timelines
[on_finish]: https://www.jspsych.org/overview/callbacks/#on_finish-experiment

**Reply by william-hopper** [reply]
Hi!

The stimuli will just be plain letters and numbers (more specifically, participant will see a stream of letters with a target number interwoven like so: "bgsIuH7nfhdfjhsN7Hfgw..." etc with the participant having to respond when they see the number 7. html would definitely work therefore but there will be no gap between stimuli i.e. every 350 ms the stimuli changes to the next one.

html-keyboard-response seemed like the logical place to start and, like you say, each stimulus being considered as a jspsych trial also seems the correct direction to head. Approaching it like this, and changing the response window to be 1050 ms (350*3), I guess the idea would be to check if the participant had responded at each timestep and then see if when they responded falls within the 3 jspsych trial range (hence changing the response window to 1050 ms). 

Also, ideally, the stimuli stream will be pseudorandomly generated at the beginning of the experiment (pseudo such that the position of the targets is controlled e.g. not within 1000 ms of each other nor the beginning or then end of the trial etc..) such that I don't have to pre-generate all the strings of stimuli. 

Perhaps it could be as simple as using a for loop to generate the timeline where each stimulus is a jspysch trial and each jspsych trial that contains the target stimulus plus the two subsequent ones have a plugin parameter like 'target_trial' = true then using the on_finish function to first check if this trial was a target trial, and if so, check if the participant responded. The RT would then be the RT given by the html-keybaord-response plugin + 350*(index of current trial - index of stimulus onset trial).

What do you think?

**Reply by vijaymarupudi** [reply]
If you know what type an target/non-target is when defining the trial, you can use the [data parameter] to tag it with the type instead of the `on_finish()` callback.

Just to clarify, so you're planning to make each stimulus its own jsPsych trial? So trials come in triplets, and by the third trial, the participant must have responded? Shouldn't the response window per trial be 350ms?

> Perhaps it could be as simple as using a for loop to generate the timeline where each stimulus is a jspysch trial and each jspsych trial that contains the target stimulus plus the two subsequent ones have a plugin parameter like 'target_trial' = true then using the on_finish function to first check if this trial was a target trial, and if so, check if the participant responded. The RT would then be the RT given by the html-keybaord-response plugin + 350*(index of current trial - index of stimulus onset trial).

This part suggests to me that you're displaying multiple stimuli in one trial. While that's definitely possible, it might be more complicated to code up. Although I do agree that a loop approach is best.

[data parameter]: https://www.jspsych.org/overview/trial/#the-data-parameter
